aws/middleware/middleware_test.go

```diff
@@ -187,3 +188,98 @@ func TestAttemptClockSkewHandler(t *testing.T) {
 		})
 	}
 }
+
+func TestRecursionDetection(t *testing.T) {
+	const lambdaFunc = "AWS_LAMBDA_FUNCTION_NAME"
+	const traceID = "_X_AMZN_TRACE_ID"
+	const traceIDHeaderKey = "X-Amzn-Trace-Id"
+
+	cases := map[string]struct {
+		LambdaFuncName string
+		TraceID        string
+		HeaderBefore   string
+		HeaderAfter    string
+	}{
+		"non lambda env and no trace ID header before": {},
+		"with lambda env but no trace ID env variable, no trace ID header before": {
+			LambdaFuncName: "some-function1",
+		},
+		"with lambda env and trace ID env variable, no trace ID header before": {
+			LambdaFuncName: "some-function2",
+			TraceID:        "traceID1",
+			HeaderAfter:    "traceID1",
+		},
+		"with lambda env and trace ID env variable, has trace ID header before": {
+			LambdaFuncName: "some-function3",
+			TraceID:        "traceID2",
+			HeaderBefore:   "traceID1",
+			HeaderAfter:    "traceID1",
+		},
+		"with lambda env and trace ID (needs encoding) env variable, no trace ID header before": {
+			LambdaFuncName: "some-function4",
+			TraceID:        "traceID3\n",
+			HeaderAfter:    "traceID3%0A",
+		},
+		"with lambda env and trace ID (contains chars must not be encoded) env variable, no trace ID header before": {
+			LambdaFuncName: "some-function5",
+			TraceID:        "traceID4-=;:+&[]{}\"'",
+			HeaderAfter:    "traceID4-=;:+&[]{}\"'",
+		},
+	}
+
+	for name, c := range cases {
+		t.Run(name, func(t *testing.T) {
+			// has to pre-check if current os has lambda function and trace ID environment variable
+			// if exists, need to restore them after each test case
+			initialLambdaFunc, hasInitialLambdaFunc := os.LookupEnv(lambdaFunc)
```

Comment:
It's unfortunate that the testing utils that already [exist](https://github.com/aws/aws-sdk-go-v2/blob/main/internal/awstesting/util.go#L97) for this depend on `aws` package. 

I think we can simplify this by just clearing the environment and restoring it after the test such that each test gets a clean environment, [example](https://github.com/aws/aws-sdk-go-v2/blob/main/aws/middleware/user_agent_test.go#L120)

Then you won't need the additional `setEnvVar`/`recordEnvVar` functions and we guarantee tests are isolated